### PR TITLE
feat: API to List all Expense Claims for given Employee

### PR DESCRIPTION
**Pull request by bhattdevarsh**
[TASK](https://app.asana.com/0/inbox/1192407897688003/1199514490117781/1199514495378656)

Added New API to List all the Expense claim created By the Given Employee

### CODE REVIEW

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 1]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
 			}
 		]
 	}).insert(ignore_permissions=True, ignore_mandatory=True)
+
+
+@frappe.whitelist()
+def list_expense_claim(user):
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
def get_expense_claims(user):
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 2]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	"""
+	API endpoint to get all Expense Claim list and its child table
+
+	User email id args required to list existing data created by that perticular user (Request From Driver App).
+	"""
+
+	employee = frappe.get_value("Employee", {"company_email": user}, "employee_name" )
+	if not employee:
+		return "Employee's E-mail ID Not Found, Please Add 'Company Email' Under 'Employee' Document"
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
		return __("Employee's E-mail ID Not Found, Please Add 'Company Email' Under 'Employee' Document.")
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 3]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+	claim_list = frappe.get_list("Expense Claim", filters={"title": employee})
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	expense_claims = frappe.get_all("Expense Claim", filters={"title": employee}, fields=["*"])
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 4]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+	expense_claim = []
+
+	if not claim_list:
+		return "No Claim Found For "+ employee
+
+	for claim in claim_list:
+		expense_claim.append(frappe.get_doc("Expense Claim", claim))
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: `
```suggestion
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py (hunk 3 above)
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: - Add filter for Employee Link field in Expense Claim and not title

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 5]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+	return expense_claim
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	return expense_claims
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/expense_claim/expense_claim.py (hunk 4 above)
Comment (bhattdevarsh) [reply erpnext/hr/doctype/expense_claim/expense_claim.py]: we need Child table as well, fields=[*] is not giving the required output so we had to iterate it through loop

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 6]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+@frappe.whitelist()
+def list_expense_claims(user):
+	"""
+	API endpoint to get all Expense Claim list and its child table
+
+	User email id args required to list existing data created by that perticular user (Request From Driver App).
+	"""
+
+	employee = frappe.get_value("Employee", {"company_email": user}, "name" )
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	employee = frappe.get_value("Employee", {"company_email": user}, "name")
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 7]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	if not employee:
+		return __("Employee's E-mail ID Not Found, Please Add 'Company Email' Under 'Employee' Document")
+
+	expense_claims = frappe.get_all("Expense Claim", filters={"employee": employee})
+
+	expense_claim_list = []
+
+	if not expense_claims:
+		return "No Claim Found For "+ employee
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
		return _("No Claim Found for {0}.").format(employee)
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 8]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	API endpoint to get all Expense Claim list and its child table
+
+	User email id args required to list existing data created by that perticular user (Request From Driver App).
+	"""
+
+	employee = frappe.get_value("Employee", {"company_email": user}, "name" )
+	if not employee:
+		return __("Employee's E-mail ID Not Found, Please Add 'Company Email' Under 'Employee' Document")
+
+	expense_claims = frappe.get_all("Expense Claim", filters={"employee": employee})
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	expense_claims = frappe.get_all("Expense Claim", filters={"employee": employee}. fields=["*"])
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 9]
```diff
@@ -405,3 +405,28 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+	expense_claim_list = []
+
+	if not expense_claims:
+		return "No Claim Found For "+ employee
+
+	for claim in expense_claims:
+		expense_claim_list.append(frappe.get_doc("Expense Claim", claim))
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: You do not need to do a `get_doc` since you can get all the fields in `get_all` itself

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 10]
```diff
@@ -405,3 +405,47 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	expense_claim_type: updated claim type, (Optional)
+	description: updated Description, (Optional)
+	amount: updated Amount, (Optional)
+	"""
+	# updated_values
+	expense_claim_details = frappe.get_doc("Expense Claim Detail", expense_claim_details_name)
+	if expense_date: expense_claim_details.update({"expense_date":expense_date})
+	if description: expense_claim_details.update({"description":description})
+	if expense_claim_type: expense_claim_details.update({"expense_claim_type":expense_claim_type})
+	if amount: expense_claim_details.update({"amount":amount})
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	if expense_date:
		expense_claim_details.update({"expense_date":expense_date})
		
	if description:
		expense_claim_details.update({"description":description})
		
	if expense_claim_type:
		expense_claim_details.update({"expense_claim_type":expense_claim_type})
		
	if amount:
		expense_claim_details.update({"amount":amount})
		
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 11]
```diff
@@ -405,3 +405,47 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	if not expense_claims:
+		return _("No Claim Found for {0}.").format(user)
+
+	return expense_claims
+
+@frappe.whitelist()
+def update_expense_claim(expense_claim_details_name, expense_date = None, expense_claim_type = None, description = None, amount = None):
+	"""
+	API endpoint to update the existing data of given Expense Claim
+	ARGS:
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion

	Args:
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 12]
```diff
@@ -405,3 +405,47 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+										"`tabExpense Claim Detail`.expense_date, `tabExpense Claim Detail`.expense_type", "`tabExpense Claim Detail`.description, `tabExpense Claim Detail`.amount,`tabExpense Claim Detail`.name"]
+								 )
+
+	if not expense_claims:
+		return _("No Claim Found for {0}.").format(user)
+
+	return expense_claims
+
+@frappe.whitelist()
+def update_expense_claim(expense_claim_details_name, expense_date = None, expense_claim_type = None, description = None, amount = None):
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
def update_expense_claim(expense_claim_detail, expense_date=None, expense_claim_type=None, description=None, amount=None):
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 13]
```diff
@@ -405,3 +405,47 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	API endpoint to update the existing data of given Expense Claim
+	ARGS:
+	expense_claim_details_name: Expense Claim Details ID to upddate perticular child data (This Data Can be found in 'list_expense_claims' api as 'name' key)
+	expense_date: updated date, (Optional)
+	expense_claim_type: updated claim type, (Optional)
+	description: updated Description, (Optional)
+	amount: updated Amount, (Optional)
+	"""
+	# updated_values
+	expense_claim_details = frappe.get_doc("Expense Claim Detail", expense_claim_details_name)
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	expense_claim_detail = frappe.get_doc("Expense Claim Detail", expense_claim_detail)
	
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 14]
```diff
@@ -405,3 +405,47 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+		return _("No Claim Found for {0}.").format(user)
+
+	return expense_claims
+
+@frappe.whitelist()
+def update_expense_claim(expense_claim_details_name, expense_date = None, expense_claim_type = None, description = None, amount = None):
+	"""
+	API endpoint to update the existing data of given Expense Claim
+	ARGS:
+	expense_claim_details_name: Expense Claim Details ID to upddate perticular child data (This Data Can be found in 'list_expense_claims' api as 'name' key)
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	expense_claim_detail: Expense Claim Details ID to upddate perticular child data (This Data Can be found in 'list_expense_claims' api as 'name' key)
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 15]
```diff
@@ -405,3 +405,51 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+	employee = frappe.get_value("Employee", {"company_email": user}, "name")
+	if not employee:
+		return _("Employee's E-mail ID Not Found, Please Add 'Company Email' Under 'Employee' Document")
+
+	expense_claims = frappe.get_all("Expense Claim",
+								filters={"employee": employee},
+								fields=["name", "employee_name", "approval_status",
+										"expense_approver", "`tabExpense Claim Detail`.*"]
+								)
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	expense_claims = frappe.get_all("Expense Claim", filters={
		"employee": employee
	}, fields=["name", "employee_name", "approval_status", "expense_approver", "`tabExpense Claim Detail`.*"])
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 16]
```diff
@@ -405,3 +405,49 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+
+@frappe.whitelist()
+def list_expense_claims(user):
+	"""
+	API endpoint to get all Expense Claim list and its child table
+
+	User email id args required to list existing data created by that perticular user (Request From Driver App).
+	"""
+
+	employee = frappe.get_value("Employee", {"company_email": user}, "name")
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]:
```suggestion
	employee = frappe.db.exists("Employee", {"company_email": user})
```

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 17]
```diff
@@ -405,3 +405,48 @@ def create_expense_claim(user, expense_date, expense_claim_type, description, am
...
+	"""
+	API endpoint to update the existing data of given Expense Claim
+	args:
+	expense_claim_detail: Expense Claim Details ID to upddate perticular child data (This Data Can be found in 'list_expense_claims' api as 'name' key)
+	expense_date: updated date, (Optional)
+	expense_claim_type: updated claim type, (Optional)
+	description: updated Description, (Optional)
+	amount: updated Amount, (Optional)
+	"""
+	# updated_values
```
Comment (hrwX) [inline erpnext/hr/doctype/expense_claim/expense_claim.py:454]:
```suggestion

```

**Review by hrwX — DISMISSED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**